### ISSUE TITLE
Netdata CPU usage and Frequency monitoring ignores Boost

### ISSUE THREAD

**Issue by Windyo** [netdata/netdata#10077]
##### Bug report summary
On the Netdata dashboard:
- CPU Utilization caps at 100%
- CPU Freq caps at the on-demand indicated max.

My CPU has Boost enabled, with Boost being able to go up to 200% of the normal load (from 1.6Ghz max  to 3.0Ghz max).
Boost is active and when checking the CPU in the machine itself by `cat /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq` I can see that the CPU does in fact scale up to 2.8Ghz under load with no difficulty.
However Netdata reports the max freq as always being 1.6Ghz, which I deem a bug.

##### OS / Environment
```
Netdata running in docker on:

/etc/lsb-release:DISTRIB_ID=Ubuntu
/etc/lsb-release:DISTRIB_RELEASE=19.10
/etc/lsb-release:DISTRIB_CODENAME=eoan
/etc/lsb-release:DISTRIB_DESCRIPTION="Ubuntu 19.10"
/etc/os-release:NAME="Ubuntu"
/etc/os-release:VERSION="19.10 (Eoan Ermine)"
/etc/os-release:ID=ubuntu
/etc/os-release:ID_LIKE=debian
/etc/os-release:PRETTY_NAME="Ubuntu 19.10"
/etc/os-release:VERSION_ID="19.10"
/etc/os-release:HOME_URL="https://www.ubuntu.com/"
/etc/os-release:SUPPORT_URL="https://help.ubuntu.com/"
/etc/os-release:BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
/etc/os-release:PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
/etc/os-release:VERSION_CODENAME=eoan
/etc/os-release:UBUNTU_CODENAME=eoan
```

##### Netdata version
netdata v1.25.0-163-gbbc15fb2

##### Component Name
CPU

##### Steps To Reproduce

1. View CPU dash
2. Put system under load
3. report on CPU Mhz using `cat /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq`
4. compare values

##### Expected behavior

CPU reporting also reports on Boost state and actual speed.

**Comment by Ferroin** [comment]
For CPU utilization, it quite simply does not work the way you seem to expect, in _any_ tool. What the percentage shown for CPU utilization actually is is a measure of what percentage of time the CPU has actually been executing machine code over the past second. The fact that it’s time-based is important for monitoring for three reasons:

* It provides consistency with network and disk utilization (the other two big bottlenecks n most systems) are tracked, making it easy to understand performance data at a glance.
* It tells you very directly whether your CPU is your bottleneck or not. If you’re doing something and it’s constantly hitting max CPU utilization, then it’s performance is limited by the CPU.
* It can be computed without needing to know anything about instruction timings, historical data about the CPU frequency, the intricate details of the CPU’s execution pipeline, and a large number of other things that are difficult or impossible to measure on a large number of systems.

Additionally, because it’s time based, it also cannot go over 100% for any one logical CPU core or 100% times the number of logical CPUs.

Notably, any other widely used tool (including other operating systems system monitors) works this way as well for tracking CPU usage.

--

On the note of CPU frequency, it’s a bit more complicated.

Strictly speaking, the listed 'max' frequency is correct. That’s what the CPU branding will say, that’s what the CPU will run at in most cases under full load, and that’s what the CPU is designed to run at under most circumstances.

Additionally though, on many systems there is not a way to reliably determine the max boost frequency, and the max boost frequency may not even be constant (on Intel systems for example, it’s dependent on RAPL configuration done by the firmware or OS, which can change depending on the system’s power source and/or power saving settings, and it may even be _lower_ than the processor’s reported max frequency).

---

Now, if you’re using the CPU frequency monitoring in Netdata, and _that_ is not correctly showing the boosted speed, that that indeed is a bug, and it’s one we would like to fix.

**Comment by Windyo** [comment]
Hey 😁
Re: utilization, noted and understood. Let's leave that by the wayside. 
-------------
Re: frequency monitoring I use the CPU frequency component in the home dashboard. I can confirm it never shows a boosted state even when I know that the rate is in fact boosted, and when querying the current freq tells me that the freq is in fact over 1.6Ghz.

**Comment by Ferroin** [comment]
OK, can you provide some more info about your specific CPU as well as the contents of `/sys/devices/system/cpu/cpufreq/policy0/scaling_driver` and `/sys/devices/system/cpu/cpufreq/policy0/scaling_governor`.

**Comment by Windyo** [comment]
CPU info:
```Processor Information
        Socket Designation: P0
        Type: Central Processor
        Family: Opteron X3000
        Manufacturer: AuthenticAMD
        ID: FF FB 8B 17 01 0F 66 00
        Signature: Family 11, Model 15, Stepping 15
        Flags:
                FPU (Floating-point unit on-chip)
                CX8 (CMPXCHG8 instruction supported)
                APIC (On-chip APIC hardware supported)
                SEP (Fast system call)
                PSE-36 (36-bit page size extension)
                PSN (Processor serial number present and enabled)
                DS (Debug store)
                ACPI (ACPI supported)
        Version: AMD Opteron(tm) X3216 APU
        Voltage: 1.0 V
        External Clock: 100 MHz
        Max Speed: 1600 MHz
        Current Speed: 1600 MHz
        Status: Populated, Enabled
        Upgrade: None
        L1 Cache Handle: 0x001B
        L2 Cache Handle: 0x001C
        L3 Cache Handle: Not Provided
        Serial Number: Not Specified
        Asset Tag: Not Specified
        Part Number: Not Specified
        Core Count: 2
        Core Enabled: 2
        Thread Count: 2
        Characteristics:
                64-bit capable```



`/sys/devices/system/cpu/cpufreq/policy0/scaling_driver`
`acpi-cpufreq`

`/sys/devices/system/cpu/cpufreq/policy0/scaling_governor`
`ondemand`


Additional data:
`cpupower frequency-info`
```
analyzing CPU 0:
  driver: acpi-cpufreq
  CPUs which run at the same hardware frequency: 0
  CPUs which need to have their frequency coordinated by software: 0
  maximum transition latency: 4.0 us
  hardware limits: 1.20 GHz - 1.60 GHz
  available frequency steps:  1.60 GHz, 1.40 GHz, 1.20 GHz
  available cpufreq governors: conservative ondemand userspace powersave performance schedutil
  current policy: frequency should be within 1.20 GHz and 1.60 GHz.
                  The governor "ondemand" may decide which speed to use
                  within this range.
  current CPU frequency: 1.20 GHz (asserted by call to hardware)
  boost state support:
    Supported: yes
    Active: yes
    Boost States: 5
    Total States: 8
    Pstate-Pb0: 3000MHz (boost state)
    Pstate-Pb1: 2800MHz (boost state)
    Pstate-Pb2: 2500MHz (boost state)
    Pstate-Pb3: 2100MHz (boost state)
    Pstate-Pb4: 1900MHz (boost state)
    Pstate-P0:  1600MHz
    Pstate-P1:  1400MHz
    Pstate-P2:  1200MHz```

let me know if I can give you anything else that could be useful

**Comment by Ferroin** [comment]
I have personally seen boost frequency reporting issues with AMD CPUs of the same vintage in the past, _but_ that was years ago on a much older version of Linux and it included the data reproted in sysfs not being consistently correct, so I don’t think that’s what you are seeing here.

Given this, I do believe you have indeed found a bug in Netdata itself. Unfortunately I don’t have AMD hardware that old to try and reproduce it (and I’m not sure if any of our other developers do either), and all the AMD systems I do have are reporting boost speeds correctly (but they’re also all Ryzen systems, so there’s enough of a difference in the CPU that I’m not inclined to say it’s not reproducible, just that I can’t personally reproduce it due to lack of hardware to do so).

---

@netdata/agent I’m not sure who should pick this up as I don’t remember who’s responsible for the cpufreq collector and do not have a solid enough knowledge of how it works to properly debug the code to figure out what exactly is going on here, so I would appreciate if someone could appropriately assign the issue I would greatly appreciate it.

**Comment by ilyam8** [comment]
@vlvkobal take a look

**Comment by Saruspete** [comment]
Hello,

From the output of `cpupower frequency-info` it's your governor that limits the max freq to 1.6GHz 

```
current policy: frequency should be within 1.20 GHz and 1.60 GHz.
```

Try by doing `cpupower frequency-set --max 3000000``` (freq on cli are in KHz)

**Comment by ilyam8** [comment]
> report on CPU Mhz using cat /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq
> compare values

🤔 i dont understand how they can be different, netdata uses same files to collect cpu freq

https://github.com/netdata/netdata/blob/767cbf022517e3ae0f05fcf882e1353e0d05ab63/collectors/proc.plugin/proc_stat.c#L542-L543

Keep in mind that freq changes a lot and doing `cat ...` you dont get exactly the same value you see on the dashboards (netdata reads the file once every second)

**Comment by vlvkobal** [comment]
I don't see any action we could do about the issue. Closing it due to inactivity.

**Comment by RX14** [comment]
I've found that for AMD cpus, the `stats/time_in_state` reports the right values, and `scaling_cur_freq` does not.

I've added a systemd unit override containing
```
[Service]
ExecStartPre=+/bin/sh -c 'chmod 700 /sys/devices/system/cpu/cpu*/cpufreq/stats/'
```
to work around this, and force netdata to fall back to the other file, and I get accurate frequency statistics.

Perhaps changing the order of these files on AMD cpus will give better results, or perhaps the latter file is good for intel CPUs too?

**Comment by ilyam8** [comment]
@tycho hi, you seem to have a very good understanding of the subject. [What do you think?](https://github.com/netdata/netdata/issues/10077#issuecomment-861905902)

**Comment by tycho** [comment]
This is one of the reasons I did the turbostat plugin before. It accounts for the turbo frequencies, most likely the same way as on Intel (through MSRs). Unfortunately my turbostat plugin doesn't work anymore because the output format of `turbostat --Dump` changed at some point. I probably just need to bite the bullet and read the MSRs from `/dev/cpu/*/msr` directly, but that would require knowing what MSRs to look at (and they can vary by CPU vendor, which is why I preferred letting turbostat do the heavy lifting since it would have theoretically been lower maintenance, but that assumption did not hold).

**Comment by Saruspete** [comment]
Could you please provide the CPU Model, kernel & OS version + the output of `cpupower frequency-info` & `scaling_cur_freq` please ?

Since 4.13.14 / 4.14.0 there was [changes in kernel code](https://bugzilla.kernel.org/show_bug.cgi?id=197009) for /proc/cpuinfo, which may involve the internal loop that updated this quick value to be not well updated.

both  [cpupower](https://github.com/torvalds/linux/tree/master/tools/power/cpupower) and [turbostat](https://github.com/torvalds/linux/tree/master/tools/power/x86/turbostat) are from the kernel, and rely mostly on aperf/mperf to get the frequency, so they should be working whatever kernel / driver is used.

That's not the case for sysfs files, where the value is pre-computed in cpufreq driver, and does not always behave correctly, depending on CPU Model, kernel version & cpufreq driver.

**Comment by RX14** [comment]
Oh, I mistyped: `stats/time_in_state` gives the wrong values (because it is limited to only the acpi frequency steps without boost), and `scaling_cur_freq` gives the correct value.

```
$ cat /proc/cpuinfo | grep 'model name' | head -n 1&& uname -srvmo && cpupower frequency-info && s grep . /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq /sys/devices/system/cpu/cpu0/cpufreq/stats/time_in_state
model name	: AMD Ryzen 7 PRO 4750U with Radeon Graphics
Linux 5.12.10-zen1-1-zen #1 ZEN SMP PREEMPT Thu, 10 Jun 2021 16:34:48 +0000 x86_64 GNU/Linux
analyzing CPU 0:
  driver: acpi-cpufreq
  CPUs which run at the same hardware frequency: 0
  CPUs which need to have their frequency coordinated by software: 0
  maximum transition latency:  Cannot determine or is not supported.
  hardware limits: 1.40 GHz - 1.70 GHz
  available frequency steps:  1.70 GHz, 1.60 GHz, 1.40 GHz
  available cpufreq governors: conservative ondemand userspace powersave performance schedutil
  current policy: frequency should be within 1.40 GHz and 1.70 GHz.
                  The governor "schedutil" may decide which speed to use
                  within this range.
  current CPU frequency: Unable to call hardware
  current CPU frequency: 1.99 GHz (asserted by call to kernel)
  boost state support:
    Supported: yes
    Active: no
/sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq:3163079
/sys/devices/system/cpu/cpu0/cpufreq/stats/time_in_state:1700000 2806406
/sys/devices/system/cpu/cpu0/cpufreq/stats/time_in_state:1600000 20112
/sys/devices/system/cpu/cpu0/cpufreq/stats/time_in_state:1400000 11386814
```

**Comment by Saruspete** [comment]
Yeah, it's the acpi_cpufreq module that just reports the BIOS populated values, and when the turbo is enabled, you have no control on it. That was the same with intel in older kernels (or if you disable the intel_pstate)
The new intel_pstate driver provides accurate freq + hints on the turbo so you can ask for min/max frequencies, even in turbo-boost mode.

We should discard the time_in_state readings when the acpi_cpufreq driver is in use.

**Comment by ilyam8** [comment]
@Saruspete is that something you would be willing to help with? 🙃

**Comment by Saruspete** [comment]
Well, I forgot about this thread, and I just encountered the issue on a server of mine...

Also, it seems the `/proc/cpuinfo` has also the frequency, but they are sometimes slightly different from scaling_cur_freq. and it seems not available on all architectures (like arm)
However, they have the benefit of always being available in a single file.

It's available on a 7763 + centos8 / kernel 4.18.0-305 + acpi_cpufreq (note: the 11 first values are different, but sometimes there's none, 4, or any other number)
```
paste  <(grep MHz /proc/cpuinfo) <(cat /sys/devices/system/cpu/cpu{0..128}/cpufreq/scaling_cur_freq 2>/dev/null)
cpu MHz		: 1794.944	2045525 !
cpu MHz		: 1664.530	2215512 !
cpu MHz		: 1790.131	2383451 !
cpu MHz		: 1791.357	2071048 !
cpu MHz		: 1791.624	1833465 !
cpu MHz		: 1791.310	1798437 !
cpu MHz		: 1791.093	2659319 !
cpu MHz		: 1795.134	2473494 !
cpu MHz		: 1795.462	1622837 !
cpu MHz		: 1793.808	3058666 !
cpu MHz		: 1795.035	1971392 !
cpu MHz		: 1791.672	1791672
cpu MHz		: 1793.894	1793894
cpu MHz		: 1791.534	1791534
cpu MHz		: 1794.491	1794491
cpu MHz		: 1790.378	1790378
cpu MHz		: 2708.025	2708025
cpu MHz		: 2950.422	2950422
cpu MHz		: 2705.029	2705029
cpu MHz		: 2982.398	2982398
cpu MHz		: 2710.680	2710680
cpu MHz		: 2927.972	2927972
cpu MHz		: 3426.615	3426615
cpu MHz		: 2733.490	2733490
cpu MHz		: 2710.227	2710227
cpu MHz		: 2717.474	2717474
cpu MHz		: 2707.711	2707711
cpu MHz		: 3341.285	3341285
cpu MHz		: 3285.684	3285684
cpu MHz		: 2729.616	2729616
cpu MHz		: 2707.695	2707695
cpu MHz		: 2722.766	2722766
cpu MHz		: 1795.589	1795589
cpu MHz		: 1792.410	1792410
cpu MHz		: 1794.580	1794580
cpu MHz		: 1794.774	1794774
cpu MHz		: 1792.885	1792885
cpu MHz		: 1793.734	1793734
cpu MHz		: 1792.782	1792782
cpu MHz		: 1793.767	1793767
cpu MHz		: 1794.783	1794783
cpu MHz		: 1792.342	1792342
cpu MHz		: 1793.098	1793098
cpu MHz		: 1793.114	1793114
cpu MHz		: 1795.554	1795554
cpu MHz		: 1792.919	1792919
cpu MHz		: 1796.186	1796186
cpu MHz		: 1796.765	1796765
cpu MHz		: 1792.138	1792138
cpu MHz		: 1793.551	1793551
cpu MHz		: 1792.914	1792914
cpu MHz		: 1792.444	1792444
cpu MHz		: 1795.370	1795370
cpu MHz		: 1794.498	1794498
cpu MHz		: 1793.990	1793990
cpu MHz		: 1796.530	1796530
cpu MHz		: 1795.037	1795037
cpu MHz		: 1794.747	1794747
cpu MHz		: 1796.193	1796193
cpu MHz		: 1796.119	1796119
cpu MHz		: 1793.912	1793912
cpu MHz		: 1793.623	1793623
cpu MHz		: 1794.411	1794411
cpu MHz		: 1793.081	1793081
cpu MHz		: 2810.865	2810865
cpu MHz		: 3464.224	3464224
cpu MHz		: 2812.545	2812545
cpu MHz		: 2812.083	2812083
cpu MHz		: 2806.047	2806047
cpu MHz		: 2812.955	2812955
cpu MHz		: 2814.206	2814206
cpu MHz		: 2808.488	2808488
cpu MHz		: 1792.396	1792396
cpu MHz		: 1792.182	1792182
cpu MHz		: 1793.719	1793719
cpu MHz		: 1791.277	1791277
cpu MHz		: 1793.403	1793403
cpu MHz		: 1674.393	1674393
cpu MHz		: 1791.417	1791417
cpu MHz		: 1793.891	1793891
cpu MHz		: 1791.924	1791924
cpu MHz		: 1793.893	1793893
cpu MHz		: 1793.427	1793427
cpu MHz		: 1795.811	1795811
cpu MHz		: 1795.902	1795902
cpu MHz		: 1791.924	1791924
cpu MHz		: 1792.923	1792923
cpu MHz		: 1792.374	1792374
cpu MHz		: 1794.127	1794127
cpu MHz		: 1794.565	1794565
cpu MHz		: 1793.475	1793475
cpu MHz		: 1793.677	1793677
cpu MHz		: 1794.583	1794583
cpu MHz		: 1794.299	1794299
cpu MHz		: 1792.657	1792657
cpu MHz		: 1792.677	1792677
cpu MHz		: 1793.360	1793360
cpu MHz		: 1792.968	1792968
cpu MHz		: 1793.100	1793100
cpu MHz		: 1794.831	1794831
cpu MHz		: 1791.488	1791488
cpu MHz		: 1794.638	1794638
cpu MHz		: 1792.720	1792720
cpu MHz		: 1791.833	1791833
cpu MHz		: 1793.560	1793560
cpu MHz		: 1793.435	1793435
cpu MHz		: 1791.525	1791525
cpu MHz		: 1791.723	1791723
cpu MHz		: 1672.832	1672832
cpu MHz		: 1791.633	1791633
cpu MHz		: 1792.476	1792476
cpu MHz		: 1793.118	1793118
cpu MHz		: 1795.041	1795041
cpu MHz		: 1790.843	1790843
cpu MHz		: 1793.941	1793941
cpu MHz		: 1792.992	1792992
cpu MHz		: 1792.356	1792356
cpu MHz		: 1793.428	1793428
cpu MHz		: 1793.802	1793802
cpu MHz		: 1790.768	1790768
cpu MHz		: 1793.726	1793726
cpu MHz		: 1792.607	1792607
cpu MHz		: 1791.235	1791235
cpu MHz		: 1793.318	1793318
cpu MHz		: 1791.930	1791930
cpu MHz		: 1792.394	1792394
cpu MHz		: 1793.676	1793676
cpu MHz		: 1795.321	1795321
```

Also working on my desktop i7 + intel_pstate + 5.4.109
having less cores, I don't see these differences between files & cpuinfo. Maybe the values are not updated while we read ?
```
$ paste  <(grep MHz /proc/cpuinfo) <(cat /sys/devices/system/cpu/cpu{0..8}/cpufreq/scaling_cur_freq 2>/dev/null)
cpu MHz		: 1327.162	1327162
cpu MHz		: 1747.786	1747786
cpu MHz		: 1527.420	1527420
cpu MHz		: 1255.674	1255674
cpu MHz		: 1516.034	1516034
cpu MHz		: 3170.642	3170642
cpu MHz		: 1481.500	1481500
cpu MHz		: 3191.666	3191666

$ paste  <(grep MHz /proc/cpuinfo) <(cat /sys/devices/system/cpu/cpu{0..8}/cpufreq/scaling_cur_freq 2>/dev/null)
cpu MHz		: 881.581	881581
cpu MHz		: 848.174	848174
cpu MHz		: 917.705	917705
cpu MHz		: 897.553	897553
cpu MHz		: 890.328	890328
cpu MHz		: 873.249	873249
cpu MHz		: 896.748	896748
cpu MHz		: 896.935	896935
```


That generates a lot of different possibilities. I would tend to allow letting change the preferred source in the config file.

**Comment by RX14** [comment]
On my ryzen laptop:
```
$ paste  <(grep MHz /proc/cpuinfo) <(cat /sys/devices/system/cpu/cpu{0..16}/cpufreq/scaling_cur_freq 2>/dev/null)
cpu MHz		: 3132.783	1827944
cpu MHz		: 1400.000	2194849
cpu MHz		: 1400.000	1861139
cpu MHz		: 1400.000	1814358
cpu MHz		: 1400.000	1822022
cpu MHz		: 1400.000	1834486
cpu MHz		: 1400.000	4100944
cpu MHz		: 4112.115	4109274
cpu MHz		: 1397.471	1429833
cpu MHz		: 1400.000	1398913
cpu MHz		: 1397.412	2185563
cpu MHz		: 1400.000	1400463
cpu MHz		: 1603.316	1682833
cpu MHz		: 1400.000	1411075
cpu MHz		: 1400.000	1413092
cpu MHz		: 1400.000	1412554
```

seems fairly different, especially where it interacts with hyperthreading. The scaling_cur_freq feels like a longer-term average than the `/proc/cpuinfo` values. Since the actual clock frequency in the chip is likely only ever instantaneously integer multiples of 100MHz because of how the PLL works.

**Comment by thulle** [comment]
@vlvkobal maybe we could reopen this issue?

$ cpupower frequency-info|grep steps
  available frequency steps:  3.40 GHz, 2.80 GHz, 2.20 GHz

3400 MHz is the max that gets logged in netdata for me, but both cpuinfo and scaling_cur_freq reports higher frequencies:

# paste  <(grep MHz /proc/cpuinfo) <(cat /sys/devices/system/cpu/cpu{0..16}/cpufreq/scaling_cur_freq 2>/dev/null)
[...]
cpu MHz		: 4625.010	4625010
cpu MHz		: 4649.976	4649976

**Comment by thulle** [comment]
Found #13116 instead.
